File: Day5/findMedianFromDataStream.java
package Day5;

// import java.util.PriorityQueue;
import java.util.*;
public class findMedianFromDataStream {
    PriorityQueue<Integer> minHeap = new PriorityQueue<>();
    PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
    
    public void addNum(int num){
        maxHeap.offer(num);
        minHeap.offer(maxHeap.poll());
        if(minHeap.size()>maxHeap.size()){
            maxHeap.offer(minHeap.poll());
        }
    }

    public double findMedian(){
        if (maxHeap.size()>minHeap.size()) return maxHeap.peek();
        return (minHeap.peek()+maxHeap.peek())/2.0d;
    }
}   

File: Day5/findMedianFromDataStream.py
from heapq import *

class MedianFinder:

    def __init__(self):
        self.minHeap=[]
        self.maxHeap=[]
    
    def addNum(self,num):
        heappush(self.maxHeap,-num)
        heappush(self.minHeap,-heappop(self.maxHeap))
        if len(self.minHeap)>len(self.maxHeap):
            heappush(self.maxHeap,-heappop(self.minHeap))
    
    def findMedian(self):
        if (len(self.maxHeap)>len(self.minHeap)):
            return -self.maxHeap[0]
        return (-self.maxHeap[0]+self.minHeap[0])/2